compiler/hash-ast/src/operator.rs

```diff
@@ -1,37 +1,201 @@
+    BitOr,
+    /// '||'
+    Or,
+    /// '&'
+    BitAnd,
+    /// '&&'
+    And,
+    /// '^'
+    BitXor,
+    /// '^^'
+    Exp,
+    /// '>'
+    Gt,
+    /// '>='
+    GtEq,
+    /// '<'
+    Lt,
+    /// '<='
+    LtEq,
+    /// '>>'
+    Shr,
+    /// '<<'
+    Shl,
+    /// '+'
+    Add,
+    /// '-'
+    Sub,
+    /// '*'
+    Mul,
+    /// '/'
+    Div,
+    /// '%'
+    Mod,
 
-// Types of unary operators that might need to be transformed.
-pub enum UnaryOpType {
-    FnCall(&'static str),
-    Ref,
-    Deref,
+    /// 'as'
+    // @@Cleanup: This is a special operator since it performs a higher order operation compared
+    //            to all other members of the Operator enum, maybe it should be in it's own enum??
+    As,
 }
 
-/// Enum representing the type of compound function that an operator represents.
-/// @@Documentation: What's the difference between LazyNamed and Named? - LazyNamed operators exhibit short circuiting behaviour.
-#[derive(Debug)]
-pub enum OperatorFn {
-    Named {
-        name: Cow<'static, str>,
-        assigning: bool,
-    },
-    LazyNamed {
-        name: Cow<'static, str>,
-        assigning: bool,
-    },
-    Compound {
-        name: CompoundFn,
-        assigning: bool,
-    },
+/// This implementation will be used for printing code from tokens.
+impl std::fmt::Display for OperatorKind {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            OperatorKind::EqEq => write!(f, "=="),
+            OperatorKind::NotEq => write!(f, "!="),
+            OperatorKind::BitOr => write!(f, "|"),
+            OperatorKind::Or => write!(f, "||"),
+            OperatorKind::BitAnd => write!(f, "&"),
+            OperatorKind::And => write!(f, "&&"),
+            OperatorKind::BitXor => write!(f, "^"),
+            OperatorKind::Exp => write!(f, "^^"),
+            OperatorKind::Gt => write!(f, ">"),
+            OperatorKind::GtEq => write!(f, ">="),
+            OperatorKind::Lt => write!(f, "<"),
+            OperatorKind::LtEq => write!(f, "<="),
+            OperatorKind::Shr => write!(f, "<<"),
+            OperatorKind::Shl => write!(f, ">>"),
+            OperatorKind::Add => write!(f, "+"),
+            OperatorKind::Sub => write!(f, "-"),
+            OperatorKind::Mul => write!(f, "*"),
+            OperatorKind::Div => write!(f, "/"),
+            OperatorKind::Mod => write!(f, "%"),
+            OperatorKind::As => write!(f, "as"),
+        }
+    }
 }
 
-/// Compound functions that use multiple function calls when they are transformed.
-#[derive(Debug)]
-pub enum CompoundFn {
-    Leq,
-    Geq,
-    Lt,
-    Gt,
+/// This implementation will be used for error messages. The debug formatter is
+/// mainly used in error reports whereas the display formatter is used when
+/// printing code from tokens.
+impl std::fmt::Debug for OperatorKind {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
```

Comment:
Isn't this the same as the to_string implementation? Maybe this should be derived instead to show the actual enum variant?

Reply:
Yeah I think this was just copied over from the old sources.